### PR TITLE
feat: Adding scanner for invariants for schedules v2

### PR DESCRIPTION
**Pull request by davidporter-id-au**
## What changed?

This introduces an optional scanner for looking through v2 schedules and to perform some invariant checks and metric checks. The theory is that:
- We should generally not see schedules with the nextActionTime in the past, and for those that do, they should be paused or be waiting for the workflow to fire. 
- We should not see workflows that are closed that are still around for a long period of time 
- We should expect (eventually, after migration and rollout of schedules-v2) all workflows to fit into the category of closed or running. 

Some of these assertions are possibly a bit premature as my understanding of the system is still ramping up, but it's relatively low risk, all the scanner does is emit a log or metric.

## Why?

The intent is to just get a better handle on the schedules feature as it rolls out, owing to its criticality. It's hard to test for all edge-cases ahead of time, so looking for invariants in production can be illuminating.

## How did you test it?
- [ ] built
- [X] run locally and tested manually
- [ ] covered by existing tests
- [X] added new unit test(s)
- [ ] added new functional test(s)

For testing, I intentionally added a bug in the scheduler to hang the schedule after firing and then let it run and added lots of loud log messages so I could see if the scanner was able to detect it, which it was. 

This does *not* cover the cases where permissions might differ across namespaces, I still need to investigate that a bit more.

## Potential risks

Probably fairly low: 
- A serious bug could cause a hot-loop of continue-as-new workflows if the activities were both misconfigured and unable to start, but otherwise it's unlikely to do much.
- A misconfiguration in the rate-limit might cause visibility to get somewhat hammered by the scanner, generating some visibility load.

### CODE REVIEW

File: service/worker/scanner/scheduleinvariants/scavenger.go [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+	"go.temporal.io/server/common/quotas"
+	"go.temporal.io/server/common/sdk"
+)
+
+// totalNamespaceTag is the value used on the cluster-wide aggregate metric emission.
+const totalNamespaceTag = "__total__"
```
Comment (davidporter-id-au) [inline service/worker/scanner/scheduleinvariants/scavenger.go]: I don't think I like this, @davidporter-id-au  to look more about how we do aggregation

**Comment by chaptersix** [comment]
nit, could rename the files?
`service/worker/scanner/scheduleinvariants/scavenger.go`

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+// ScanStuckOpen is a long-running activity that periodically scans for schedules with
+// ExecutionStatus=Completed and CloseTime older than the configured buffer.
+func (a *Activities) ScanStuckOpen(ctx context.Context) error {
+	return a.runForeverWithInterval(ctx, func(scanCtx context.Context) error {
+		threshold := a.timeSource.Now().UTC().Add(scheduler.DefaultTweakables.IdleTime * scheduleIdleTimeBufferMultiplier).Format(time.RFC3339Nano)
+		query := fmt.Sprintf(`ExecutionStatus = "Completed" AND CloseTime < "%s"`, threshold)
+		return a.runScan(scanCtx, "stuck_open", query, metrics.ScheduleInvariantsScannerStuckOpenCount.Name())
+	})
+}
```
Comment (chaptersix) [inline service/worker/scanner/scheduleinvariants/invariants.go:127]: I don't think this is a valid check. A completed schedule is completed just like a workflow is completed. 

While a schedule is Idle, it remains running.

File: service/worker/scanner/scheduleinvariants/workflows.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+// StuckOpenWorkflow scans for schedules that appear stuck open long after their CloseTime.
+func StuckOpenWorkflow(ctx workflow.Context) error {
+	return runScanActivity(ctx, StuckOpenActivityName, StuckOpenWorkflow)
+}
+
+// UnknownStateWorkflow scans for running, unpaused schedules with no
+// TemporalScheduleNextActionTime.
+func UnknownStateWorkflow(ctx workflow.Context) error {
+	return runScanActivity(ctx, UnknownStateActivityName, UnknownStateWorkflow)
+}
```
Comment (chaptersix) [inline service/worker/scanner/scheduleinvariants/workflows.go:85]: it's valid for a schedule to not have a next action time if it has a scheduled idle task

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+			tag.WorkflowNamespace(nsName), tag.NewStringTag("schedule_id", scheduleID), tag.Error(err))
+		return false
+	}
+
+	if desc.GetSchedule().GetState().GetPaused() {
+		return true
+	}
+	overlap := desc.GetSchedule().GetPolicies().GetOverlapPolicy()
+	if (overlap == enumspb.SCHEDULE_OVERLAP_POLICY_BUFFER_ONE || overlap == enumspb.SCHEDULE_OVERLAP_POLICY_BUFFER_ALL) &&
+		len(desc.GetInfo().GetRunningWorkflows()) > 0 {
```
Comment (davidporter-id-au) [inline service/worker/scanner/scheduleinvariants/invariants.go:386]: my understanding of these policies is pretty poor, so just calling this as being worthy of scruitiny

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 2 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:127]: Thank you for that, hopefully addressed

**Comment by davidporter-id-au** [comment]
> nit, could rename the files? `service/worker/scanner/scheduleinvariants/scavenger.go`

Sorry, not quite following what they should be renamed to?

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+		// the query's intention is:
+		// show running schedules where ScheduleNextActionTime is in the past (plus a small buffer)
+		query := fmt.Sprintf(
+			`%s < "%s" AND TemporalSchedulePaused = false AND ExecutionStatus = "Running"`,
+			scheduler.ScheduleNextActionTimeName,
+			threshold,
+		)
+		return a.runOverdueScan(scanCtx, query)
+	})
+}
```
Comment (chaptersix) [inline service/worker/scanner/scheduleinvariants/invariants.go:105]: just to confirm. Will NextActionTime be null if there are no future actions?

File: service/worker/scanner/scheduleinvariants/scavenger.go [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+// ScanStuckOpen is a long-running activity that periodically scans for schedules whose
+// ScheduleIdleCloseTime deadline has already passed (beyond the configured buffer).
+// ScheduleIdleCloseTime is only emitted while a schedule is open, so an overdue value
+// means the idle-close task never fired and the schedule is stuck open. Paused schedules
+// are excluded since they are intentionally held open and do not idle-close.
+func (a *Activities) ScanStuckOpen(ctx context.Context) error {
+	return a.runForeverWithInterval(ctx, func(scanCtx context.Context) error {
+		threshold := a.timeSource.Now().UTC().Add(-(scheduler.DefaultTweakables.IdleTime * scheduleIdleTimeBufferMultiplier)).Format(time.RFC3339Nano)
+		query := fmt.Sprintf(
+			`%s < "%s" AND TemporalSchedulePaused = false`,
```
Comment (chaptersix) [inline service/worker/scanner/scheduleinvariants/scavenger.go]: you'll need this on probably all of these queries:
` AND ExecutionStatus = "Running"``

Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/scavenger.go]: Good catch. Confusingly this is not actually a valid parameter for the Cli, but I believe it's ok in V2 world. 
```
$ temporal workflow list --namespace  some-ns.a2dd6 --query 'ExecutionStatus = "Running"'
  Status                       WorkflowId                               Type             StartTime
  Running  wf-id  QaAgent                 42 minutes ago
  Running  session-manager                                     SessionManagerWorkflow  1 week ago
  Running  document-lookup-temporal-docs-v2                    DocumentLookup          2 weeks ago
$ temporal schedule list --namespace  some-ns.a2dd6 --query 'ExecutionStatus = "Running"'
time=2026-06-05T00:31:43.240 level=ERROR msg="invalid query filter for schedules: cannot filter on \"ExecutionStatus\""
```

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 5 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:105]: Yes, it's removed. Tested locally

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+
+// ScanStuckOpen is a long-running activity that periodically scans for schedules whose
+// ScheduleIdleCloseTime deadline has already passed (beyond the configured buffer).
+// ScheduleIdleCloseTime is only emitted while a schedule is open, so an overdue value
+// means the idle-close task never fired and the schedule is stuck open. Paused schedules
+// are excluded since they are intentionally held open and do not idle-close.
```
Comment (chaptersix) [inline service/worker/scanner/scheduleinvariants/invariants.go:111]: Bug: this adds `IdleTime * 2` to `Now()`, producing a threshold ~14 days in the **future**. The query `CloseTime < "<future>"` will match virtually every completed schedule, not just stuck ones. Should negate the duration:
```go
threshold := a.timeSource.Now().UTC().Add(-scheduler.DefaultTweakables.IdleTime * scheduleIdleTimeBufferMultiplier)
```

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+	// activity. It must be comfortably below the activity's HeartbeatTimeout.
+	heartbeatInterval = 10 * time.Second
+	// scheduleListPageSize controls pagination of ListChasmExecutions inside
+	// ScanOverdueNextActionTime. Each page is also one unit of rate-limited progress.
+	scheduleListPageSize             = 100
+	scheduleIdleTimeBufferMultiplier = 2
+	namespaceListPageSize            = 100
+)
+
+// Activities holds shared dependencies for all three schedule-invariants scanner activities.
```
Comment (chaptersix) [inline service/worker/scanner/scheduleinvariants/invariants.go:37]: `overdueScheduleCap` is declared but never referenced. Either wire it into `runOverdueScan`/`forEachScheduleInNamespace` to cap results, or remove it.

**Review by chaptersix — COMMENTED** [review]
Review comments below.

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+		}
+		resp, err := a.visibilityManager.ListChasmExecutions(ctx, &visibilityservice.ListChasmExecutionsRequest{
+			ArchetypeId:   chasm.SchedulerArchetypeID,
+			NamespaceId:   nsID.String(),
+			Namespace:     nsName,
+			Query:         query,
+			PageSize:      scheduleListPageSize,
+			NextPageToken: pageToken,
+		})
+		if err != nil {
```
Comment (chaptersix) [inline service/worker/scanner/scheduleinvariants/invariants.go:348]: Use `tag.ScheduleID(scheduleID)` instead of `tag.NewStringTag("schedule_id", ...)` -- the helper exists and uses the canonical key `"schedule-id"`. Same on the other Warn log at line 205.

**Comment by chaptersix** [comment]
> > nit, could rename the files? `service/worker/scanner/scheduleinvariants/scavenger.go`
> 
> Sorry, not quite following what they should be renamed to?

`scavenger.go` just seems like it might have been left over from something else. not something I'll block the PR on if you think the name fits

**Review by chaptersix — APPROVED** [review]
LGTM after the fixes above

File: service/worker/scanner/scheduleinvariants/scavenger.go [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+// be a false positive for up to the retention window, and genuinely stuck-open schedules
+// (past ScheduleIdleCloseTime) would be double-counted with ScanStuckOpen.
+func (a *Activities) ScanUnknownState(ctx context.Context) error {
+	return a.runForeverWithInterval(ctx, func(scanCtx context.Context) error {
+		query := fmt.Sprintf(
+			`%s IS NULL AND %s IS NULL AND TemporalSchedulePaused = false AND ExecutionStatus = "Running"`,
+			scheduler.ScheduleNextActionTimeName,
+			scheduler.ScheduleIdleCloseTimeName,
+		)
+		a.logger.Info("schedule-invariants: starting unknown_state scan",
```
Comment (lina-temporal) [inline service/worker/scanner/scheduleinvariants/scavenger.go]: (minor) should we add similar logging for the others?

File: service/worker/scanner/scheduleinvariants/scavenger.go [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+	"go.temporal.io/server/common/quotas"
+	"go.temporal.io/server/common/sdk"
+)
+
+const (
+	// scanInterval is how often each scanner activity kicks off a fresh scan pass.
+	scanInterval = 15 * time.Minute
```
Comment (lina-temporal) [inline service/worker/scanner/scheduleinvariants/scavenger.go]: This interval might be nice to have as dynamic config.

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 7 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:111]: I'm not sure that's accurate? The line on 108 looks two weeks in the past and uses it to select for schedules which are idled from before then I think (? I'm easily confused, happy to be corrected)

File: service/worker/scanner/scanner.go [hunk 12]
```diff
@@ -209,6 +220,66 @@ func (s *Scanner) Start() error {
...
+			work.RegisterActivityWithOptions(scheduleActivities.ScanStuckOpen, activity.RegisterOptions{Name: scheduleinvariants.StuckOpenActivityName})
+
+			if err := work.Start(); err != nil {
+				return err
+			}
+		}
+
+		if s.context.cfg.ScheduleInvariantsScannerUnknownStateEnabled() {
+			s.wg.Add(1)
+			go s.startWorkflowWithRetry(ctx, scheduleinvariants.UnknownStateWFStartOptions, scheduleinvariants.UnknownStateWorkflowName)
```
Comment (lina-temporal) [inline service/worker/scanner/scanner.go:275]: What do you think of a single workflow where logic on which scans to execute lives within the activity? I suggest it since the activities/workflows appear to be long-running, indefinite things anways.

**Review by lina-temporal — APPROVED** [review]

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 8 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:37]: I think possibly this is stale?

File: service/worker/scanner/scheduleinvariants/scavenger.go (hunk 10 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/scavenger.go]: let me remove that actually, it's noise

File: service/worker/scanner/scanner.go (hunk 12 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scanner.go:275]: I honestly don't feel too strongly, if that feels conceptually cleaner / better, I don't see any harm not to, other inertia.

Comment (davidporter-id-au) [reply service/worker/scanner/scanner.go:275]: Let me come back to this if it's possible if it's of significant value, weakly I'd prefer to keep as-is just to allow for individual enablement, though this isn't a particularly persuasive justification

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 7 above)
Comment (chaptersix) [reply service/worker/scanner/scheduleinvariants/invariants.go:111]: my local changes where out of date. disregard

File: common/dynamicconfig/constants.go [hunk 13]
```diff
@@ -604,6 +604,51 @@ is currently processing a task.
 		`BuildIdScavengerVisibilityRPS is the rate limit for visibility calls from the build id scavenger`,
 	)
 
+	ScheduleInvariantsScannerOverdueNextActionTimeEnabled = NewGlobalBoolSetting(
```
Comment (dnr) [inline common/dynamicconfig/constants.go:607]: If these are all generally used around the same time, what do you think about making one struct-valued config instead?

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+	}
+
+	for _, nsName := range namespaces {
+		var nsAnomalies int64
+		err := a.forEachScheduleInNamespace(ctx, nsName, query, func(scheduleID string) {
+			if a.scheduleIsExpectedNotToFire(ctx, nsName, scheduleID) {
+				return
+			}
+			nsAnomalies++
+			a.logger.Warn("schedule-invariants - the schedule was expected to fire but doesn't appear to have done so",
```
Comment (dnr) [inline service/worker/scanner/scheduleinvariants/invariants.go:234]: the operation/component scope is already attached to the logger?
```suggestion
			a.logger.Warn("schedule was expected to fire but doesn't appear to have done so",
```

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 15]
```diff
@@ -0,0 +1,411 @@
...
+	return nil
+}
+
+// ListAllNamespaces returns the names of every namespace that exists in the cluster
+// metadata store and is active in the current cluster. Pagination is handled internally.
+func (a *Activities) ListAllNamespaces(ctx context.Context) ([]string, error) {
+	var names []string
+	var pageToken []byte
+	for {
+		nsResp, err := a.metadataManager.ListNamespaces(ctx, &persistence.ListNamespacesRequest{
```
Comment (dnr) [inline service/worker/scanner/scheduleinvariants/invariants.go:253]: can't we just use namespaceRegistry.GetAllNamespaces()? we're requiring them to be in the registry below anyway

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+}
+
+// forEachScheduleInNamespace paginates ListChasmExecutions for the given namespace
+// (scoped to the Scheduler archetype) and invokes visit for every schedule entity in
+// the result. The rate limiter is consumed per page.
+func (a *Activities) forEachScheduleInNamespace(
+	ctx context.Context,
+	nsName string,
+	query string,
+	visit func(scheduleID string),
```
Comment (dnr) [inline service/worker/scanner/scheduleinvariants/invariants.go:325]: it could be fun (+  more readable) to make this a go iterator since it's already most of the way there

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 17]
```diff
@@ -0,0 +1,411 @@
...
+// visibility is actually in a state that legitimately explains the stale NextActionTime:
+// either the schedule is paused, or it's holding under BUFFER_ONE / BUFFER_ALL with a
+// prior workflow still running.
+func (a *Activities) scheduleIsExpectedNotToFire(ctx context.Context, nsName, scheduleID string) bool {
+	if err := a.rateLimiter.Wait(ctx); err != nil {
+		a.logger.Warn("rate-limiter wait failed during DescribeSchedule; counting schedule as anomaly",
+			tag.WorkflowNamespace(nsName), tag.ScheduleID(scheduleID), tag.Error(err))
+		return false
+	}
+	desc, err := a.sdkClientFactory.GetSystemClient().WorkflowService().DescribeSchedule(ctx, &workflowservice.DescribeScheduleRequest{
```
Comment (dnr) [inline service/worker/scanner/scheduleinvariants/invariants.go:371]: is it possible to get what you need from the memo returned from visibility? that'd be much cheaper

File: service/worker/scanner/scheduleinvariants/workflows.go [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/sdk/workflow"
+)
+
+const (
+	OverdueNextActionTimeWorkflowName = "schedule-invariants-overdue-next-action-time-scanner"
+	OverdueNextActionTimeActivityName = "scan-schedule-invariants-overdue-next-action-time"
```
Comment (dnr) [inline service/worker/scanner/scheduleinvariants/workflows.go:14]: I'm usually all about consistency, but this pattern of using these string names instead of just the functions themselves bothers me so much when I'm navigating go workflow code... what do you think about breaking the trend and just using well-named functions in RegisterWorkflow/RegisterActivity?

actually you already use the functions instead of the string alises in in workflow.NewContinueAsNewError, it's that natural... and I think that won't even work if you register them using strings

File: service/worker/scanner/scanner.go [hunk 19]
```diff
@@ -209,6 +222,68 @@ func (s *Scanner) Start() error {
 		}
 	}
 
+	scheduleInvariantsAnyEnabled := s.context.cfg.ScheduleInvariantsScannerOverdueNextActionTimeEnabled() ||
+		s.context.cfg.ScheduleInvariantsScannerStuckOpenEnabled() ||
+		s.context.cfg.ScheduleInvariantsScannerUnknownStateEnabled()
+	if scheduleInvariantsAnyEnabled && !s.context.visibilityManager.HasStoreName(elasticsearch.PersistenceName) {
+		s.context.logger.Info("schedule-invariants scanners are enabled but advanced (Elasticsearch) visibility is not configured; skipping")
```
Comment (dnr) [inline service/worker/scanner/scanner.go:229]: there is no "advanced" visibility anymore, all features are at parity (afaik)

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 16 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:325]: oh, this is fancy, I might have spend a bit of time to get my head around their spec

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 17 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:371]: My honest answer is I have no idea, let me take a look at the data there and see what it is. I'd be a little surprised if we were specifying the number of running workflows there though

File: service/worker/scanner/scanner.go (hunk 19 above)
Comment (rodrigozhou) [reply service/worker/scanner/scanner.go:229]: As @dnr said, no need to do `!s.context.visibilityManager.HasStoreName(elasticsearch.PersistenceName)`.

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 20]
```diff
@@ -0,0 +1,411 @@
...
+		}
+	}
+}
+
+// runScan ties the per-scanner pieces together: list namespaces, fan out a single
+// visibility query per namespace, and emit per-namespace metrics. Cluster-wide
+// totals are computed by the metrics backend by summing across the namespace tag.
+// Used by the count-only scanners (StuckOpen, UnknownState).
+func (a *Activities) runScan(ctx context.Context, subScanner, query, metricName string) error {
+	namespaces, err := a.ListAllNamespaces(ctx)
```
Comment (rodrigozhou) [inline service/worker/scanner/scheduleinvariants/invariants.go:197]: How often are you running these scans? Should we consider caching the list of namespaces over querying persistence in every scan? Given you have the namespace registry, you can register a callback whenever a namespace is changed/deleted to invalidate items from the cache.

File: service/worker/scanner/scheduleinvariants/invariants.go [hunk 21]
```diff
@@ -0,0 +1,411 @@
...
+
+// runOverdueScan lists individual matching schedules per namespace, calls
+// DescribeSchedule on each, and filters out schedules that are paused or
+// expected to not be able to fire (BUFFER_ONE / BUFFER_ALL waiting on a prior
+// running workflow). What remains is counted as an anomaly.
+func (a *Activities) runOverdueScan(ctx context.Context, query string) error {
+	const subScanner = "overdue_next_action_time"
+	metricName := metrics.ScheduleInvariantsScannerOverdueNextActionTimeCount.Name()
+
+	namespaces, err := a.ListAllNamespaces(ctx)
```
Comment (rodrigozhou) [inline service/worker/scanner/scheduleinvariants/invariants.go:222]: ditto

**Review by rodrigozhou — APPROVED** [review]

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 20 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:197]: every 15 minutes, but as per David's point, there's no reason not to use the NS cache, that's just a miss on me

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 15 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:253]: Addressed in https://github.com/temporalio/temporal/pull/10596

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 16 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:325]: Addressed in https://github.com/temporalio/temporal/pull/10596

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 17 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:371]: After a bit more of a think and look, I don't want to rely on the memo. A stuck or stale scheduler will definitionally not be updating the associated visibility record, whereas if I look to see the actual Execution state (for what is expected to be zero-to-a-very-small amount of workflows) to see if they're genuinely waiting for some reason, then I'll catch this case. 

I think it's a good point about cost though. I've added a cap of 100 executions. If it finds more than that it'll just bail out. This is maybe imperfect from a detection perspective, but it's better than self-DOSing

File: service/worker/scanner/scheduleinvariants/workflows.go (hunk 18 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/workflows.go:14]: addressed in https://github.com/temporalio/temporal/pull/10596

File: service/worker/scanner/scanner.go (hunk 19 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scanner.go:229]: Addressed in https://github.com/temporalio/temporal/pull/10596

File: service/worker/scanner/scheduleinvariants/invariants.go (hunk 21 above)
Comment (davidporter-id-au) [reply service/worker/scanner/scheduleinvariants/invariants.go:222]: Addressed in https://github.com/temporalio/temporal/pull/10596
